## Code Review - Claude Code (Multi-Perspective)

> 🔒 보안/권한 · ⚙️ 운영/성능 · 📐 패턴/일관성 — 3개 관점 병렬 리뷰 종합

### Summary
| Severity | Count |
|----------|-------|
| Blocker  | 0     |
| Critical | 3     |
| Major    | 4     |
| Minor    | 5     |

---

### Critical

**[CR-001] 🔒⚙️📐 port-forward.sh: 백그라운드 프로세스 정리(trap) 누락**
- 파일: `scripts/port-forward.sh` (line 13-18)
- 현재: `kubectl port-forward` 3개를 `&`로 실행 후 `wait`만 수행. Ctrl+C 시 백그라운드 프로세스가 orphan으로 남아 포트 점유 → 재실행 시 `bind: address already in use` 에러 발생.
- 수정 제안: `trap`으로 cleanup 함수 추가.
```bash
cleanup() {
  echo ""
  echo "포트포워딩 종료..."
  kill $(jobs -p) 2>/dev/null
  wait 2>/dev/null
}
trap cleanup EXIT INT TERM
```

**[CR-002] 📐 setup.sh: `cd` 후 원래 디렉토리로 복귀하지 않음**
- 파일: `scripts/setup.sh` (line 32)
- 현재: `cd "${ROOT_DIR}/infrastructure/argocd"` 후 복귀 없음. 이후 단계는 절대 경로라 현재는 동작하지만, 신규 단계 추가 시 상대 경로 명령이 실패하는 함정.
- 수정 제안: subshell `( cd ... && helm ... )` 또는 `pushd/popd` 사용.

**[CR-003] 🔒 setup.sh: ArgoCD 초기 비밀번호 터미널 stdout 평문 출력**
- 파일: `scripts/setup.sh` (line 53-57)
- 현재: base64 디코딩한 비밀번호를 `echo`로 출력. CI 로그, 터미널 히스토리, 화면 공유 시 노출 위험.
- 수정 제안: 직접 출력 대신 조회 명령어 안내만 출력.
```bash
echo "  Password: 아래 명령으로 확인"
echo "    kubectl -n argocd get secret argocd-initial-admin-secret -o jsonpath='{.data.password}' | base64 -d"
```

---

### Major

**[CR-004] 📐 변수 인용부호(quoting) 불일치 — shellcheck SC2086**
- 파일: `scripts/setup.sh` (line 23, 34, 38, 42)
- 현재: `${ARGOCD_NAMESPACE}` 참조에 쌍따옴표가 있는 곳과 없는 곳 혼재.
- 수정 제안: 모든 변수 참조를 `"${VAR}"` 형태로 통일.

**[CR-005] 📐⚙️ Makefile CLUSTER_NAME 미사용 + 스크립트와 DRY 위반**
- 파일: `Makefile` (line 3), `scripts/setup.sh` (line 6), `scripts/teardown.sh` (line 4)
- 현재: Makefile에 `CLUSTER_NAME := goti-dev` 선언했지만 참조 없음. 각 스크립트가 독립적으로 하드코딩.
- 수정 제안: (A) Makefile에서 `export CLUSTER_NAME`하고 스크립트에서 `${CLUSTER_NAME:-goti-dev}`로 받기, 또는 (B) Makefile에서 미사용 변수 제거.

**[CR-006] 🔒📐⚙️ ArgoCD URL 프로토콜 불일치 (http vs https)**
- 파일: `scripts/port-forward.sh` (line 5) vs `scripts/setup.sh` (line 54) vs `README.md` (line 32)
- 현재: port-forward.sh만 `https://`, 나머지는 `http://`. insecure mode이므로 `http://`가 정확.
- 수정 제안: port-forward.sh line 5를 `http://localhost:30080`으로 통일.

**[CR-007] ⚙️ Makefile validate 타겟의 `head -20` 출력 제한**
- 파일: `Makefile` (line 38, 42)
- 현재: `| head -20`으로 검증 결과 잘림. 20줄 이후의 에러를 놓칠 수 있고, SIGPIPE로 앞 프로세스 exit code가 비정상 가능.
- 수정 제안: `head -20` 제거하여 전체 출력 표시.

---

### Minor

**[CR-008] ⚙️ setup.sh: Helm `--wait` + kubectl wait 중복 대기**
- 파일: `scripts/setup.sh` (line 36-43)
- 현재: Helm `--wait --timeout 5m` 성공 시 deployment는 이미 available → kubectl wait는 즉시 통과 (중복). Helm 실패 시 set -e로 중단되어 kubectl wait에 도달 불가.
- 수정 제안: 의도적이라면 주석 명시, 아니면 하나 제거.

**[CR-009] 🔒⚙️ teardown.sh: `--force` 플래그 미지원**
- 파일: `scripts/teardown.sh` (line 7-10)
- 현재: `read -p`로만 확인. CI/non-interactive 환경에서 항상 취소됨.
- 수정 제안: `--force` 인자 지원 검토 (dev 환경 우선순위 낮음).

**[CR-010] 📐 setup.sh: 에러 메시지가 stderr가 아닌 stdout으로 출력**
- 파일: `scripts/setup.sh` (line 55)
- 수정 제안: `|| echo "비밀번호를 가져올 수 없습니다" >&2`

**[CR-011] ⚙️ Makefile lint가 goti-server만 대상**
- 파일: `Makefile` (line 31)
- 현재: 단일 차트이므로 문제없음. MSA 전환 시 charts/ 순회 로직 검토 필요.

**[CR-012] 🔒 Makefile clean의 `find -exec rm -rf` 범위**
- 파일: `Makefile` (line 45-46)
- 수정 제안: `-maxdepth 3` 추가 또는 `charts/*/charts` 패턴으로 한정하면 더 안전.

---

### Good Practices

- **🔒📐 `set -euo pipefail` 일관 적용** — 모든 스크립트에 적용되어 에러 즉시 중단.
- **⚙️ Kind 클러스터 생성 멱등성** — 존재 여부 확인 후 건너뛰기 처리.
- **📐 namespace 생성 멱등성** — `--dry-run=client -o yaml | kubectl apply -f -` 패턴.
- **⚙️ teardown 삭제 전 확인 프롬프트** — 실수 방지 안전장치.
- **📐 Makefile self-documenting help** — `.DEFAULT_GOAL := help` + grep/awk 자동 문서화.
- **📐 README 구조 체계적** — 전제조건 → 빠른 시작 → 접속 정보 → 디렉토리 구조 → 트러블슈팅.

--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,7 @@
 .PHONY: up down port-forward lint validate status clean help
 
-CLUSTER_NAME := goti-dev
+export CLUSTER_NAME ?= goti-dev
+export ARGOCD_NAMESPACE ?= argocd
 SHELL := /bin/bash
 
 ## 클러스터 관리
@@ -18,7 +19,7 @@ status: ## 클러스터 및 ArgoCD 상태 확인
 	@kubectl get nodes -o wide 2>/dev/null || echo "클러스터에 연결할 수 없습니다"
 	@echo ""
 	@echo "=== ArgoCD Applications ==="
-	@kubectl get applications -n argocd 2>/dev/null || echo "ArgoCD가 설치되지 않았습니다"
+	@kubectl get applications -n $(ARGOCD_NAMESPACE) 2>/dev/null || echo "ArgoCD가 설치되지 않았습니다"
 	@echo ""
 	@echo "=== Istio 상태 ==="
 	@kubectl get pods -n istio-system 2>/dev/null || echo "Istio가 설치되지 않았습니다"
@@ -35,15 +36,15 @@ validate: ## Helm template 렌더링 + dry-run 검증
 	helm dependency build charts/goti-server/
 	helm template goti-server charts/goti-server/ \
 		-f environments/dev/goti-server/values.yaml \
-		--namespace goti | kubectl apply --dry-run=client -f - 2>&1 | head -20
+		--namespace goti | kubectl apply --dry-run=client -f -
 	@echo ""
 	@echo "=== 부트스트랩 매니페스트 검증 ==="
-	kubectl apply --dry-run=client -f clusters/dev/bootstrap/ 2>&1 | head -20
+	kubectl apply --dry-run=client -f clusters/dev/bootstrap/
 
 ## 유틸리티
 clean: ## Helm dependency cache 정리
-	find charts/ -name "charts" -type d -exec rm -rf {} + 2>/dev/null || true
-	find charts/ -name "*.tgz" -delete 2>/dev/null || true
+	find charts/ -maxdepth 3 -name "charts" -type d -exec rm -rf {} + 2>/dev/null || true
+	find charts/ -maxdepth 3 -name "*.tgz" -delete 2>/dev/null || true
 
 help: ## 도움말 표시
 	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | \

--- a/scripts/port-forward.sh
+++ b/scripts/port-forward.sh
@@ -1,10 +1,19 @@
 #!/usr/bin/env bash
 set -euo pipefail
 
+# 백그라운드 프로세스 정리
+cleanup() {
+  echo ""
+  echo "포트포워딩 종료..."
+  kill $(jobs -p) 2>/dev/null || true
+  wait 2>/dev/null || true
+}
+trap cleanup EXIT INT TERM
+
 echo "=== 포트포워딩 시작 ==="
-echo "ArgoCD:     https://localhost:30080 (NodePort, 포트포워딩 불필요)"
-echo "Grafana:    http://localhost:3000"
-echo "Prometheus: http://localhost:9090"
+echo "ArgoCD:      http://localhost:30080 (NodePort, 포트포워딩 불필요)"
+echo "Grafana:     http://localhost:3000"
+echo "Prometheus:  http://localhost:9090"
 echo "goti-server: http://localhost:8080"
 echo ""
 echo "종료: Ctrl+C"

--- a/scripts/setup.sh
+++ b/scripts/setup.sh
@@ -3,13 +3,13 @@ set -euo pipefail
 
 SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 ROOT_DIR="$(dirname "$SCRIPT_DIR")"
-CLUSTER_NAME="goti-dev"
-ARGOCD_NAMESPACE="argocd"
+CLUSTER_NAME="${CLUSTER_NAME:-goti-dev}"
+ARGOCD_NAMESPACE="${ARGOCD_NAMESPACE:-argocd}"
 
 echo "=== Goti Dev 환경 셋업 시작 ==="
 
 # 1. Kind 클러스터 생성
-echo "[1/5] Kind 클러스터 생성..."
+echo "[1/4] Kind 클러스터 생성..."
 if kind get clusters 2>/dev/null | grep -q "^${CLUSTER_NAME}$"; then
   echo "  클러스터 '${CLUSTER_NAME}'가 이미 존재합니다. 건너뜁니다."
 else
@@ -17,10 +17,10 @@ else
   echo "  클러스터 생성 완료"
 fi
 
-# 2. Istio namespace 생성
-echo "[2/5] 네임스페이스 생성..."
+# 2. 네임스페이스 생성
+echo "[2/4] 네임스페이스 생성..."
 kubectl create namespace istio-system --dry-run=client -o yaml | kubectl apply -f -
-kubectl create namespace ${ARGOCD_NAMESPACE} --dry-run=client -o yaml | kubectl apply -f -
+kubectl create namespace "${ARGOCD_NAMESPACE}" --dry-run=client -o yaml | kubectl apply -f -
 kubectl create namespace goti --dry-run=client -o yaml | kubectl apply -f -
 kubectl create namespace monitoring --dry-run=client -o yaml | kubectl apply -f -
 
@@ -28,33 +28,30 @@ kubectl create namespace monitoring --dry-run=client -o yaml | kubectl apply -f 
 kubectl label namespace goti istio-injection=enabled --overwrite
 
 # 3. ArgoCD Helm 설치
-echo "[3/5] ArgoCD 설치..."
-cd "${ROOT_DIR}/infrastructure/argocd"
-helm dependency build
-helm upgrade --install argocd . \
-  -n ${ARGOCD_NAMESPACE} \
-  -f values.yaml \
-  -f values-dev.yaml \
-  --wait --timeout 5m
+echo "[3/4] ArgoCD 설치..."
+(
+  cd "${ROOT_DIR}/infrastructure/argocd"
+  helm dependency build
+  helm upgrade --install argocd . \
+    -n "${ARGOCD_NAMESPACE}" \
+    -f values.yaml \
+    -f values-dev.yaml \
+    --wait --timeout 5m
+)
 
-# 4. ArgoCD 준비 대기
-echo "[4/5] ArgoCD 준비 대기..."
-kubectl wait --for=condition=available deployment/argocd-server \
-  -n ${ARGOCD_NAMESPACE} --timeout=300s
-
-# 5. 부트스트랩 적용
-echo "[5/5] 부트스트랩 매니페스트 적용..."
+# 4. 부트스트랩 적용
+echo "[4/4] 부트스트랩 매니페스트 적용..."
 kubectl apply -f "${ROOT_DIR}/clusters/dev/bootstrap/"
 
-# ArgoCD 초기 비밀번호 출력
+# ArgoCD 접속 정보 출력
 echo ""
 echo "=== 셋업 완료! ==="
 echo ""
 echo "ArgoCD 접속 정보:"
 echo "  URL: http://localhost:30080"
-ARGOCD_PASSWORD=$(kubectl -n ${ARGOCD_NAMESPACE} get secret argocd-initial-admin-secret -o jsonpath="{.data.password}" 2>/dev/null | base64 -d || echo "비밀번호를 가져올 수 없습니다")
 echo "  Username: admin"
-echo "  Password: ${ARGOCD_PASSWORD}"
+echo "  Password: 아래 명령으로 확인"
+echo "    kubectl -n ${ARGOCD_NAMESPACE} get secret argocd-initial-admin-secret -o jsonpath='{.data.password}' | base64 -d"
 echo ""
 echo "포트포워딩: make port-forward"
 echo "상태 확인: make status"

--- a/scripts/teardown.sh
+++ b/scripts/teardown.sh
@@ -1,10 +1,16 @@
 #!/usr/bin/env bash
 set -euo pipefail
 
-CLUSTER_NAME="goti-dev"
+CLUSTER_NAME="${CLUSTER_NAME:-goti-dev}"
 
 echo "=== Goti Dev 환경 삭제 ==="
-read -p "정말 '${CLUSTER_NAME}' 클러스터를 삭제하시겠습니까? (y/N): " confirm
+
+if [[ "${1:-}" == "--force" || "${1:-}" == "-y" ]]; then
+  confirm="y"
+else
+  read -p "정말 '${CLUSTER_NAME}' 클러스터를 삭제하시겠습니까? (y/N): " confirm
+fi
+
 if [[ "$confirm" != "y" && "$confirm" != "Y" ]]; then
   echo "취소되었습니다."
   exit 0

